### PR TITLE
Fix issue with merge process and output labels (Merged and MergedError)

### LINKED ISSUE #10592: T0 related bug when creating merge process
**Impact of the bug**
 `merge` creation process  fails with T0 workflows due to a bug in the output module labels

**Describe the bug**
When creating a `merge` process, the output module label is either: `Merged` OR `MergedError`, depending on the use of `--useErrorDataset` in this external script:

https://github.com/cms-sw/cmssw-wm-tools/blob/30b626d030b7b11f83fe4e0385ce303e83d3fcdf/bin/cmssw_wm_create_process.py#L45-L46

However, the output modules listed in cmsswStep have both, so one of them fails and the process aborts.

**How to reproduce it**
T0 workflow with `useErrorDataset=False` on a `merge` process gives:

```
2021-06-09 06:47:49,472:INFO:SetupCMSSWPset:ScramRun command args: edm_pset_tweak.py --input_pkl /srv/job/WMTaskSpace/cmsRun1/PSet.pkl --output_pkl /srv/job/WMTaskSpace/cmsRun1/PSet.pkl --json /srv/job/WMTaskSpace/cmsRun1/PSetTweak.json
2021-06-09 06:47:49,472:INFO:Scram:Creating a subprocess to run the PSet setup.
2021-06-09 06:47:49,472:INFO:Scram:Also recording SCRAM command-line related output.
2021-06-09 06:47:49,481:INFO:Scram:    Invoking command: edm_pset_tweak.py --input_pkl /srv/job/WMTaskSpace/cmsRun1/PSet.pkl --output_pkl /srv/job/WMTaskSpace/cmsRun1/PSet.pkl --json /srv/job/WMTaskSpace/cmsRun1/PSetTweak.json
2021-06-09 06:47:49,884:INFO:Scram:Subprocess stdout was:
Set attribute process.source.fileNames to cms.untracked.vstring('/store/unmerged/data/Tier0_REPLAY_2021/Cosmics/RAW/v2106090600/000/341/607/00000/aa0b3ef3-6a94-4e03-a338-d4779e41f25c.root')
Set attribute process.Merged.logicalFileName to cms.untracked.string('/store/backfill/1/data/Tier0_REPLAY_2021/Cosmics/RAW/v2106090600/000/341/607/00000/Merged.root')
Set attribute process.maxEvents to cms.untracked.PSet(
    input = cms.untracked.int32(-1),
    output = cms.optional.untracked.allowed(cms.int32,cms.PSet)
)
Tweak not applied process.MergedError.logicalFileName /store/backfill/1/data/Tier0_REPLAY_2021/Cosmics-Error/RAW/v2106090600/000/341/607/00000/MergedError.root
```

**Expected behavior**
Tweak should only be created and applied to either `Merged` or `MergedError`, depending on the value of `useErrorDataset`

**Additional context and error message**
 In the past, this error was ignored when creating the data set and the process creation continued.

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #10592

#### Status
not-tested

#### Description
This solution avoids tweaking both `Merged` and `MergedError` output labels, but only one of them, depending on whether `--useErrorDataset` was used or not when creating the process.

#### Is it backward compatible (if not, which system it affects?)
YES

### CODE REVIEW

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 1]
```diff
@@ -748,6 +748,18 @@ def __call__(self):
         cmsswStep = self.step.getTypeHelper()
         for om in cmsswStep.listOutputModules():
             mod = cmsswStep.getOutputModule(om)
+            modName = str(getattr(mod, "_internal_name"))
```
Comment (amaltaro) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py:751]: Kenyi, how about defining a new "public" method in the ConfigSection object (src/python/WMCore/Configuration.py), to retrieve the object name? Something like:
```
def getInternalName(self):
    return self._internal_name
```

The default value is None, so whoever is getting it, needs to take that into consideration.

This `str()` casting here might become a problem as well, if the config section name is a byte string, it would result in `'b"name"'`.

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 2]
```diff
@@ -748,6 +748,18 @@ def __call__(self):
...
+
+            if funcName == 'merge':
+                # Do not use both Merged output label unless useErrorDataset is False
+                # Do not use both MergedError output label unless useErrorDataset is True 
+                useErrorDataset = getattr(self.jobBag, "useErrorDataset", False)
+
+                if modName.startswith('Merged'):
```
Comment (amaltaro) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py:758]: If the module name is either `Merged` or `MergedError`, then I would suggest to make a full comparison, something like:
```
if useErrorDataset and modName != 'MergedError':
    continue
if not useErrorDataset and modName == 'MergedError':
    continue
```

The `if` statement in the line above could go away as well.

**Review by amaltaro — COMMENTED** [review]
@khurtado Kenyi, the logic looks good to me, but I have made some minor suggestions in the code for your consideration. Thanks

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py (hunk 1 above)
Comment (khurtado) [reply src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py:751]: Good point. I called it that way because it's how it was used in WMTweak:
https://github.com/dmwm/WMCore/blob/psettweak/src/python/PSetTweaks/WMTweak.py#L519

But having a public method and modify both would make it more python3 friendly, thinking about the str & bytes thing.

**Comment by khurtado** [comment]
@amaltaro I just made the requested changes. Let me know if you have any other request or question or I can squash the commits otherwise ;)

**Comment by khurtado** [comment]
Update: This was tested and it worked.
There were additional errors after this, but nothing that seems related to PSet tweak changed (logcollect errors trying to write to CASTOR), which is described hr: #10110
@amaltaro Any objections in merging this ?

File: src/python/PSetTweaks/PSetTweak.py [hunk 3]
```diff
@@ -421,6 +421,16 @@ def unpersist(self, filename, formatting=None):
             for param, value in viewitems(jsoniser.parameters):
                 self.addParameter(param , value)
 
+    def clear(self):
```
Comment (amaltaro) [inline src/python/PSetTweaks/PSetTweak.py:424]: Given that this method deletes the old tweak object and recreates a default one, shall we rename it to `reset` instead?

File: src/python/WMCore/Configuration.py [hunk 4]
```diff
@@ -357,6 +357,15 @@ def listSections_(self):
         comps = self._internal_settings
         return list(comps)
 
+    def getInternalName(self):
```
Comment (amaltaro) [inline src/python/WMCore/Configuration.py:360]: Can you please make a simple unit test for this new method? Please keep it in a separate commit.

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 5]
```diff
@@ -233,8 +233,8 @@ def applyPsetTweak(self, psetTweak, skipIfSet=False, allowFailedTweaks=False, na
             cmd += " --allow_failed_tweaks"
         self.scramRun(cmd)
 
-        if cleanupTweak is True:
-            psetTweak = PSetTweak()
+        if cleanupTweak:
```
Comment (amaltaro) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py:236]: Kenyi, could you please create a docstring for this method and clarify what all these options are? For instance, I do not even know why we would want to reset the PSet tweak :) I'm sure this will help Stefano as well.

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 6]
```diff
@@ -748,8 +748,20 @@ def __call__(self):
...
+                useErrorDataset = getattr(self.jobBag, "useErrorDataset", False)
+
+                if useErrorDataset and modName != 'MergedError':
+                    continue
+                if not useErrorDataset and modName == 'MergedError':
+                    continue
+
             makeOutputTweak(mod, self.job, self.tweak)
-        self.applyPsetTweak(self.tweak, cleanupTweak=True)
+        self.applyPsetTweak(self.tweak, allowFailedTweaks=True, cleanupTweak=True)
```
Comment (amaltaro) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py:764]: A short comment above this line explaining why we allow failed tweaks to be skipped here and why we clean it up would be useful as well.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Sorry for the late review, Kenyi.
I have made some other comments along the code that are probably worth looking into.

A general comment I have is, we should make sure these changes are as clear and documented as possible. It was a very large change, and I'm the first one to be afraid of touching it when time comes ;)

**Comment by khurtado** [comment]
@amaltaro I made the changes (waiting for jenkins to finish)

**Review by amaltaro — APPROVED** [review]
Thanks for applying these final changes, Kenyi. It looks good to me, can you please just squash the first 2 commits? Thanks

**Comment by khurtado** [comment]
@amaltaro Alright! Done.

**Comment by amaltaro** [comment]
Thank you very much for the prompt fixes, Kenyi.
